### PR TITLE
feat(update-server): advertise robotModel

### PR DESCRIPTION
**Pull request by sfoster1**
The robot model is being advertised by the robot server's health
endpoint, which is good but not sufficient to robustly identify a
machine to a client across all its possible connection states and
styles.

This PR adds the robot model to
- The update server health endpoint
- a TXT record in the MDNS advertisement

This should allow identification of the machine model to be robust
across machines that are found on MDNS but unreachable; machines that
are connected with IPs with no MDNS advertisement; and everything in
between.

It does create a possible conflict (though the machine type is static
data that should not conflict). Resolution mechanisms in clients should
follow these rules:

- When more than one source (i.e. of update-server health, robot-server
health, and TXT record) is presenting a valid machine type, prioritize
in the order of 1) robot server; 2) update server; 3) MDNS
- If any source is not present (i.e. requests to `/health` or `/server/health` fail, or the MDNS advertisement is not found); or a source does not have the appropriate key (robot_model for
robot-server health, robotModel for
update-server health, the TXT record robotModel for MDNS), that source
should be disregarded
- If no source is present with a valid model entry, the robot may be assumed to be an OT-2 for now, though in the future this should be an explicit unknown-model state.

Closes RCORE-66

## Testing
- Run this on an OT-2 and OT-3
- Each should appropriately report its machine type, including on mdns
   -  it's hard to see this in MDNS; osx dns-sd can probably do it but I don't know how. You can use the wireshark query `ip.addr == 192.168.1.153 and udp.port eq 5353` (where you replace 192.168.1.153 with the ip of the robot you're testing) and look for a packet like this: `236	8.135402	192.168.1.153	224.0.0.251	MDNS	193	Standard query response 0x0000 TXT, cache flush PTR OliverT._http._tcp.local SRV, cache flush 0 0 31950 4849db.local PTR _http._tcp.local` and expand the response section to see the content

### CODE REVIEW

File: update-server/otupdate/common/constants.py [hunk 1]
```diff
@@ -19,3 +21,6 @@
 #:
 #: This ID only changes when the whole OT-2 operating system reboots.
 #: It doesn't change if some internal process merely crashes and restarts.
+
+MODEL_OT2: Final[str] = "OT-2 Standard"
+MODEL_OT3: Final[str] = "OT-3 Standard"
```
Comment (SyntaxColoring) [inline update-server/otupdate/common/constants.py:26]: What does "Standard" mean here? Why not just "OT-2"/"OT-3"?

File: update-server/otupdate/common/name_management/avahi.py [hunk 2]
```diff
@@ -312,7 +312,12 @@ def start_advertising(self, service_name: str) -> None:
             domainname,  # sdomain (.local)
             f"{hostname}.{domainname}",  # shost (hostname.local)
             dbus.UInt16(31950),  # port
-            dbus.Array([], signature="ay"),
+            # supplementary txt record with encoded machine type. this is
+            # a dbus byte array rather than a dbus string for some reason
+            # so it has to be manually encoded.
+            dbus.Array(
+                [f"robotModel={machine_type}".encode("UTF-8")], signature="ay"
+            ),  # TXT records
```
Comment (SyntaxColoring) [inline update-server/otupdate/common/name_management/avahi.py:320]: I *think* the reason it's a byte array is that [the value can actually be an arbitrary binary blob](http://www.zeroconf.org/Rendezvous/txtrecords.html#:~:text=Rules%20for%20values%20in%20DNS%2DSD%20Name/Value%20pairs).

File: update-server/otupdate/common/name_management/name_synchronizer.py [hunk 3]
```diff
@@ -41,14 +41,17 @@ class NameSynchronizer:
...
         """For internal use by this class only. Use `start()` instead."""
         self._avahi_client = avahi_client
+        self._machine_type = machine_type
 
     @classmethod
     @asynccontextmanager
     async def start(
-        cls, avahi_client: Optional[AvahiClient] = None
+        cls,
+        machine_type: str,
```
Comment (SyntaxColoring) [inline update-server/otupdate/common/name_management/name_synchronizer.py:53]: Would you mind adding the new `machine_type` arg to this method's docstring?

**Review by SyntaxColoring — APPROVED** [review]
Tested on an OT-2. Appears to be working as intended according to [this graphical DNS-SD browser](https://apps.apple.com/us/app/discovery-dns-sd-browser/id1381004916?mt=12).

Screen Shot 2022-08-05 at 11 52 02 AM

-----

Code looks good to me to merge, with some minor notes if you feel like doing them in this PR.

-----

> Resolution mechanisms in clients should follow these rules:
> 
> * When more than one source (i.e. of update-server health, robot-server
>   health, and TXT record) is presenting a valid machine type, prioritize
>   in the order of 1) robot server; 2) update server; 3) MDNS
> * If any source does not have the appropriate key (robot_model for
>   robot-server health, robotModel for
>   update-server health, the TXT record robotModel for MDNS), that source
>   may be assumed to be an OT-2
> * If any source is not present, i.e. requests to /health or
>   /server/health fail or the MDNS advertisement is not found, it should be disregarded

Note that in a worst-case failure mode, I *think* these rules can cause an OT-3 to be misidentified as an OT-2:

* mDNS and DNS-SD are unavailable; `discovery-client` is polling the robot at a static IP address.
* The robot's HTTP responses from both `update-server` and `robot-server` are unusable; perhaps they're returning `500` errors.
* Despite the unusable HTTP responses, `discovery-client` still thinks the robot is present—it's just unusable. I *think* this is a possible state if `discovery-client` successfully establishes an HTTP connection but gets an HTTP-level error code. I could be wrong.

If I'm right about this, I suggest modifying the rules to be this:

* When more than one source (i.e. of update-server health, robot-server health, and TXT record) is presenting a valid machine type, prioritize in the order of 1) robot server; 2) update server; 3) MDNS
* If any source does not have the appropriate key (robot_model for robot-server health, robotModel for update-server health, the TXT record robotModel for MDNS), ~~that source may be assumed to be an OT-2~~ **ignore that source in favor of whichever other sources are presenting valid machine types.**
* If any source is not present, i.e. requests to /health or /server/health fail or the MDNS advertisement is not found, it should be disregarded
* **If *no* source is presenting a valid machine type, the machine type is unknown. How to present this to the user is up to the client; the client can show some default model, or explicitly say the model is unknown.**

File: update-server/otupdate/common/constants.py (hunk 1 above)
Comment (sfoster1) [reply update-server/otupdate/common/constants.py:26]: it matches the values in the deck defs and allows a little more room for variation i guess. not really wedded to it

**Comment by sfoster1** [comment]
> Tested on an OT-2. Appears to be working as intended according to [this graphical DNS-SD browser](https://apps.apple.com/us/app/discovery-dns-sd-browser/id1381004916?mt=12).
> 
> Screen Shot 2022-08-05 at 11 52 02 AM
> 
> Code looks good to me to merge, with some minor notes if you feel like doing them in this PR.
> 
> > Resolution mechanisms in clients should follow these rules:
> > 
> > * When more than one source (i.e. of update-server health, robot-server
> >   health, and TXT record) is presenting a valid machine type, prioritize
> >   in the order of 1) robot server; 2) update server; 3) MDNS
> > * If any source does not have the appropriate key (robot_model for
> >   robot-server health, robotModel for
> >   update-server health, the TXT record robotModel for MDNS), that source
> >   may be assumed to be an OT-2
> > * If any source is not present, i.e. requests to /health or
> >   /server/health fail or the MDNS advertisement is not found, it should be disregarded
> 
> Note that in a worst-case failure mode, I _think_ these rules can cause an OT-3 to be misidentified as an OT-2:
> 
>     * mDNS and DNS-SD are unavailable; `discovery-client` is polling the robot at a static IP address.
> 
>     * The robot's HTTP responses from both `update-server` and `robot-server` are unusable; perhaps they're returning `500` errors.
> 
>     * Despite the unusable HTTP responses, `discovery-client` still thinks the robot is present—it's just unusable. I _think_ this is a possible state if `discovery-client` successfully establishes an HTTP connection but gets an HTTP-level error code. I could be wrong.
> 
> 
> If I'm right about this, I suggest modifying the rules to be this:
> 
>     * When more than one source (i.e. of update-server health, robot-server health, and TXT record) is presenting a valid machine type, prioritize in the order of 1) robot server; 2) update server; 3) MDNS
> 
>     * If any source does not have the appropriate key (robot_model for robot-server health, robotModel for update-server health, the TXT record robotModel for MDNS), ~that source may be assumed to be an OT-2~ **ignore that source in favor of whichever other sources are presenting valid machine types.**
> 
>     * If any source is not present, i.e. requests to /health or /server/health fail or the MDNS advertisement is not found, it should be disregarded
> 
>     * **If _no_ source is presenting a valid machine type, the machine type is unknown. How to present this to the user is up to the client; the client can show some default model, or explicitly say the model is unknown.**

I agree with pretty much all of this (and may have actually implemented it that way in the discovery client). Your misidentification case is definitely possible, but I'm not really concerned about it in the short term; it would have to be a very specific way to fail that I'm not sure is even a thing - knowing a robot is there but not caching literally any data about it.

**Pull request merged**